tools/app.hpp

```diff
@@ -0,0 +1,184 @@
+
+        [[nodiscard]] auto index_basename() const -> std::string const & { return m_basename; }
+        [[nodiscard]] auto index_encoding() const -> std::string const & { return m_encoding; }
+        [[nodiscard]] auto wand_data_path() const -> std::optional<std::string> const &
+        {
+            return m_wand_data_path;
+        }
+        [[nodiscard]] auto is_wand_compressed() const -> bool { return m_wand_compressed; }
+
+       private:
+        std::string m_basename;
+        std::string m_encoding;
+        std::optional<std::string> m_wand_data_path;
+        bool m_wand_compressed = false;
+    };
+
+    enum class QueryMode : bool { Ranked, Unranked };
+
+    template <QueryMode Mode = QueryMode::Ranked, int DefaultK = 1'000>
+    struct Query {
+        explicit Query(CLI::App *app)
+        {
+            app->add_option("-q,--queries", m_query_file, "Path to file with queries", false);
+            app->add_option(
+                "--stopwords", m_stop_words, "List of blacklisted stop words to filter out");
+            auto *terms_opt = app->add_option("--terms", m_term_lexicon, "Term lexicon");
+            app->add_option("--stemmer", m_stemmer, "Stemmer type")->needs(terms_opt);
+
+            if constexpr (Mode == QueryMode::Ranked) {
+                app->add_option("-k", m_k, "The number of top results to return", true);
+            }
+        }
+
+        [[nodiscard]] auto query_file() -> std::optional<std::reference_wrapper<std::string const>>
+        {
+            if (m_query_file) {
+                return m_query_file.value();
+            }
+            return std::nullopt;
+        }
+
+        [[nodiscard]] auto queries() const -> std::vector<::pisa::Query>
+        {
+            std::vector<::pisa::Query> q;
+            auto parse_query = resolve_query_parser(q, m_term_lexicon, m_stop_words, m_stemmer);
+            if (m_query_file) {
+                std::ifstream is(*m_query_file);
+                io::for_each_line(is, parse_query);
+            } else {
+                io::for_each_line(std::cin, parse_query);
+            }
+            return q;
+        }
+
+        [[nodiscard]] auto k() const -> int { return m_k; }
+
+       private:
+        std::optional<std::string> m_query_file;
+        int m_k = DefaultK;
+        std::optional<std::string> m_stop_words{std::nullopt};
+        std::optional<std::string> m_stemmer{std::nullopt};
+        std::optional<std::string> m_term_lexicon{std::nullopt};
+    };
+
+    struct Algorithm {
+        explicit Algorithm(CLI::App *app)
+        {
+            app->add_option("-a,--algorithm", m_algorithm, "Query processing algorithm")
+                ->required();
+        }
+
+        [[nodiscard]] auto algorithm() const -> std::string const & { return m_algorithm; }
+
+       private:
+        std::string m_algorithm;
+    };
+
+    enum class Required : bool { Required, Optional };
+
+    template <Required required = Required::Required>
```

Comment:
Why the scorer can be optional? I think it should be always required.

Reply:
I was probably thinking of when it's precomputed. But I guess we'd still specify a scorer. I'll get rid of it.